ospf6d/ospf6_area.c

```diff
@@ -848,69 +848,81 @@ DEFUN (no_area_export_list,
 	return CMD_SUCCESS;
 }
 
-DEFUN (show_ipv6_ospf6_spf_tree,
-       show_ipv6_ospf6_spf_tree_cmd,
-       "show ipv6 ospf6 spf tree",
-       SHOW_STR
-       IP6_STR
-       OSPF6_STR
-       "Shortest Path First calculation\n"
-       "Show SPF tree\n")
+DEFUN(show_ipv6_ospf6_spf_tree, show_ipv6_ospf6_spf_tree_cmd,
+      "show ipv6 ospf6 [vrf <NAME|all>] spf tree",
+      SHOW_STR IP6_STR OSPF6_STR VRF_CMD_HELP_STR
+      "All VRFs\n"
+      "Shortest Path First calculation\n"
+      "Show SPF tree\n")
 {
-	struct listnode *node;
+	struct listnode *node, *nnode;
 	struct ospf6_area *oa;
 	struct ospf6_vertex *root;
 	struct ospf6_route *route;
 	struct prefix prefix;
 	struct ospf6 *ospf6;
-
-	ospf6 = ospf6_lookup_by_vrf_name(VRF_DEFAULT_NAME);
-	OSPF6_CMD_CHECK_RUNNING(ospf6);
-	ospf6_linkstate_prefix(ospf6->router_id, htonl(0), &prefix);
-
-	for (ALL_LIST_ELEMENTS_RO(ospf6->area_list, node, oa)) {
-		route = ospf6_route_lookup(&prefix, oa->spf_table);
-		if (route == NULL) {
-			vty_out(vty, "LS entry for root not found in area %s\n",
-				oa->name);
-			continue;
+	const char *vrf_name = VRF_DEFAULT_NAME;
+	bool all_vrf = false;
+	int idx_vrf = 0;
+
+	OSPF6_FIND_VRF_ARGS(argv, argc, idx_vrf, vrf_name, all_vrf);
+
+	if (vrf_name) {
+		if (all_vrf) {
+			for (ALL_LIST_ELEMENTS_RO(om6->ospf6, nnode, ospf6)) {
+				ospf6_linkstate_prefix(ospf6->router_id,
+						       htonl(0), &prefix);
+				for (ALL_LIST_ELEMENTS_RO(ospf6->area_list,
+							  node, oa)) {
+					route = ospf6_route_lookup(
+						&prefix, oa->spf_table);
+					if (route == NULL) {
+						vty_out(vty,
+							"LS entry for root not found in area %s\n",
+							oa->name);
+						continue;
+					}
+					root = (struct ospf6_vertex *)
+						       route->route_option;
+					ospf6_spf_display_subtree(vty, "", 0,
+								  root);
+				}
+			}
+			return CMD_SUCCESS;
+		}
+		ospf6 = ospf6_lookup_by_vrf_name(vrf_name);
+		OSPF6_CMD_CHECK_RUNNING(ospf6);
+		ospf6_linkstate_prefix(ospf6->router_id, htonl(0), &prefix);
```

Comment:
The block of code starting here is identical to the one starting at line 873. Request you eliminate identical code by carving out a separate function. Alternatively the changes in this block can be accommodated in the same loop for 'all_vrf'.

Reply:
Resolved.